src/v/cluster/data_policy_manager.cc

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cluster/data_policy_manager.h"
+
+#include "cluster/cluster_utils.h"
+#include "cluster/errc.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+
+#include <system_error>
+
+namespace cluster {
+
+namespace {
+
+template<typename Cmd>
+ss::future<std::error_code> do_apply(
+  ss::shard_id shard,
+  Cmd cmd,
+  ss::sharded<data_policy_manager::container_type>& db) {
+    return db.invoke_on(
+      shard,
+      [cmd = std::move(cmd)](
+        data_policy_manager::container_type& local_db) mutable {
+          return ss::visit(
+            std::move(cmd),
+            [&local_db](create_data_policy_cmd cmd) {
+                local_db.insert_or_assign(cmd.key, std::move(cmd.value.dp));
+                return std::error_code(errc::success);
+            },
+            [&local_db](delete_data_policy_cmd cmd) {
+                if (local_db.erase(cmd.key) == 0) {
+                    return std::error_code(errc::data_policy_not_exists);
+                } else {
+                    return std::error_code(errc::success);
+                }
+            });
+      });
+}
+
+template<typename Cmd>
+ss::future<std::error_code> dispatch_updates_to_cores(
+  Cmd cmd, ss::sharded<data_policy_manager::container_type>& db) {
+    using ret_t = std::vector<std::error_code>;
+    return ss::do_with(
+      ret_t{}, [cmd = std::move(cmd), &db](ret_t& ret) mutable {
+          ret.reserve(ss::smp::count);
+          return ss::parallel_for_each(
+                   boost::irange(0, (int)ss::smp::count),
+                   [&ret, cmd = std::move(cmd), &db](int shard) mutable {
+                       return do_apply(shard, cmd, db)
+                         .then([&ret](std::error_code r) { ret.push_back(r); });
+                   })
+            .then([&ret] { return std::move(ret); })
+            .then([](std::vector<std::error_code> results) mutable {
```

Comment:
it looks like the second continuation is still within the `do_with` block so it already has access to `ret`. did you intend to chain it after `do_with`, because as written the identify `.then([&ret] { return std::move(ret); })` doesn't seem to be needed.